### PR TITLE
fix(revit/rhino): handles exception when converting null meshes in revit

### LINKED ISSUE #586: Feat (revit): support face hosted family instances
image
image

Sample files: https://drive.google.com/drive/folders/1QAMO6ytdYEzQpPkhQoqmrk77LRX0TY8c?usp=sharing

Reporter: Marco (Poldo)

### LINKED ISSUE #2311: bug: Face Based Instances
## Description

I tried different scenarios with face based family instances.

- When they are placed on a Level
- When they are placed on a System Family’s (Floor) Face
- When user hosts a face based family to another face based family’s face
- When above happens even more (face-based-ception)
- Nested Family instances (when there is a hosted family in another family)

## Issues

Especially hosting a face-based family instance on another family instance’s face (face-based-ception) is problematic.

### 2 Level Face-based-ception

In this case, there is a `Floor` that hosts `Sideboard`on its top face. `Sideboard` has `3 vases` on its top face.

Model structure:

`Floor` > `Sideboard` > `Vase`

They are not viewed correctly in the viewer nor received in Rhino.

image

Model in Revit

image

Same model in our Viewer

image

Same model in Rhino

### Face-based-ception

In this case, there are a lot of face-based hosting going on. There is a `Floor`, on its top face `Sideboard` is hosted. On `Sideboard`’s top face, `Cheese Board` hosted. On its top face, `Marble Tray` hosted. On its top face, `Coaster` hosted. On its top face `Candle Holder` is hosted.

Model Structure:

`Floor` > `Sideboard` > `Cheese Board` > `Marble Tray` > `Marble Coaster` > `Candle Holder`

They are not viewed correctly in the viewer nor received in Rhino.

image

Model in Revit

image

Same model in the viewer

image

Same model in Rhino

## Commits Used

https://latest.speckle.dev/streams/e258b0e8db/commits/1aa2777a3e

## Files Used

G:\Shared drives\All Company\07 Sample Models\Revit\Github Attachments\Instancing\Face Based Instances.rvt

### LINKED ISSUE #2316: bug: Receiving Point Based Instances (Mirrored Elements)
## Description

I tried receiving the commits send in  [Sending Point-based instances](https://www.notion.so/Sending-Point-based-instances-f84954190242432299797d38a1885089) test. I used the `revit-rhino/claire/instance-fixes `branch

Receiving in Rhino and Sketchup worked just fine. But receiving back in Revit, is problematic. 

### Receiving back in Revit

When I receive sent commit back in Revit (with relevant families loaded), only mirrored instances face correct direction. Original (non-mirrored) instances are facing the wrong direction.

image

Original model

image

After receiving in Revit

## Files Used

**Original File**

G:\Shared drives\All Company\07 Sample Models\Revit\Github Attachments\Instancing\Point Based Instances.rvt

**Receiving File**

G:\Shared drives\All Company\07 Sample Models\Revit\Github Attachments\Instancing\empty_Point Based Instances.rvt

## Streams Used

[https://latest.speckle.dev/streams/e258b0e8db/commits/b2049d6dfc](https://latest.speckle.dev/streams/e258b0e8db/commits/b2049d6dfc)

### PR DESCRIPTION
**Pull request by clairekuang**
## Description & motivation

The reported bug in #2311 was originally caused by an unhandled null exception when retrieving the display value meshes from revit elements, resulting in some definitions not having meshes in their display value which looked like missing instances.

This fix introduces a new `GetElementDisplayValue()` method for retrieving the display meshes for any revit element, which is refactor to replace many redundant methods in MeshUtils. It is currently implemented for `RevitInstance` and `Wall` conversions but should be rolled out to as many other conversions as possible (see #2314 )

This also _**improves how we were previously handling hosted and mirrored family instances**_, with better support for receiving hosted and mirrored instances (still some edge cases to be debugged here)

Also includes a minor change in rhino to cache previously converted block definitions for a slight performance gain.

Fixes #2311 
Fixes #2316 even tho this wasn't a regression
Fixes #586

## Changes:

Revit and Rhino converter

## Screenshots:

Before:

image

After:

image

### CODE REVIEW

**Comment by JR-Morgan** [comment]
Just dumping some comments here. Likely unrelated to this PR

The existing `GetFabricationMeshes` appears to fail with the default structural sample
image

And I'm noticing some foundation family instances in the wrong place
Revit foundation wrong place

File: Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertDirectShape.cs [hunk 1]
```diff
@@ -194,13 +194,15 @@ public ApplicationObject DirectShapeToNative(ApplicationObject appObj, IList<Mes
       return appObj;
     }
 
+    /* TODO: no references - deprecate?
```
Comment (AlanRynne) [inline Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertDirectShape.cs:197]: If it's not being used let's kick it.

File: Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertMeshUtils.cs [hunk 2]
```diff
@@ -11,6 +11,210 @@ namespace Objects.Converter.Revit
...
+        }
+      }
+
+      return meshBuildHelper.GetAllValidMeshes();
+    }
+
+    /// <summary>
+    /// Helper class for a single <see cref="Objects.Geometry.Mesh"/> object for each <see cref="DB.Material"/>
+    /// </summary>
+    private class MeshBuildHelper
```
Comment (AlanRynne) [inline Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertMeshUtils.cs:135]: Could we refactor this into it's own file? Or is it tied up into the Converter?

**Review by AlanRynne — CHANGES_REQUESTED** [review]
Nothing critical, only the `MeshBuildHelper` comment bellow.

File: Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertMeshUtils.cs (hunk 2 above)
Comment (AlanRynne) [reply Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertMeshUtils.cs]: Moved into separate file by myself

**Comment by AlanRynne** [comment]
Awaiting @JR-Morgan's comments/review on this.

**Review by AlanRynne — APPROVED** [review]
New changes look good. I'll wait till you confirm it's ready before merging though ;)

**Pull request merged**